File: Task/Comma-quibbling/Haskell/comma-quibbling-2.hs
import Data.List (intercalate)

--------------------- COMMA QUIBBLING --------------------

quibble :: [String] -> String
quibble ws@(_ : _ : _) =
  intercalate
    " and "
    ( [intercalate ", " . reverse . tail, head]
        <*> [reverse ws]
    )
quibble xs = concat xs

--------------------------- TEST -------------------------
main :: IO ()
main =
  mapM_ (putStrLn . (`intercalate` ["{", "}"]) . quibble) $
    [[], ["ABC"], ["ABC", "DEF"], ["ABC", "DEF", "G", "H"]]
      <> ( words
             <$> [ "One two three four",
                   "Me myself I",
                   "Jack Jill",
                   "Loner"
                 ]
         )

File: Task/Comma-quibbling/F-Sharp/comma-quibbling-2.fs
> quibble [];;
val it : string = "{}"
> quibble ["ABC"];;
val it : string = "{ABC}"
> quibble ["ABC"; "DEF"];;
val it : string = "{ABC and DEF}"
> quibble ["ABC"; "DEF"; "G"];;
val it : string = "{ABC, DEF and G}"
> quibble ["ABC"; "DEF"; "G"; "H"];;
val it : string = "{ABC, DEF, G and H}"
